src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlInternalConnectionTds.cs

```diff
@@ -2440,9 +2440,15 @@ internal SqlFedAuthToken GetFedAuthToken(SqlFedAuthInfo fedAuthInfo)
                         }
                         else
                         {
+                            SqlClientEventSource.Log.TryTraceEvent("<sc.SqlInternalConnectionTds.GetFedAuthToken.MsalServiceException error:> Timeout: {0}", serviceException.ErrorCode);
                             break;
```

Comment:
There's another potential point here for GerFedAuthToken to return a null token. If we've timed out in a cycle of 429 errors, this break will exit the loop without populating the token. We should throw an exception instead of breaking. Something along the lines of "Connection timed out while retrieving an access token. Last error: {serviceException.StatusCode}: {serviceException.Message}".